src/fairness/print_hierarchy/print_hierarchy.cpp

```diff
@@ -0,0 +1,201 @@
+
+      return 1;
+  }
+  exit = sqlite3_step(b_select_sub_banks_stmt);
+
+  // vector of tuples to store usernames and their respective shares
+  std::vector<std::tuple<std::string, std::string>> users;
+
+  // vector of strings to hold sub banks
+  std::vector<std::string> banks;
+
+  // we've reached a bank with no sub banks, so print
+  // out all associations under this sub bank
+  if (exit != SQLITE_ROW) {
+    exit = sqlite3_bind_text(b_select_associations_stmt, 1, bank_name.c_str(), -1, NULL);
+    if (exit != SQLITE_OK) {
+      fprintf(stderr, "Failed to fetch data: %s\n", sqlite3_errmsg(DB));
+      sqlite3_close(DB);
+
+      return 1;
+    }
+
+    // execute SQL statement
+    exit = sqlite3_step(b_select_associations_stmt);
+    while (exit == SQLITE_ROW) {
+      std::string user = reinterpret_cast<char const*> (sqlite3_column_text(b_select_associations_stmt, 0));
+      std::string single_user_shares = reinterpret_cast<char const*> (sqlite3_column_text(b_select_associations_stmt, 1));
+      // place the user and their allocated shares into vector
+      users.emplace_back(user, single_user_shares);
+      exit = sqlite3_step(b_select_associations_stmt);
+    }
+    // iterate through associations and append them to the hierarchy string
+    for (int i=0; i < users.size(); i++) {
+      std::string username = std::get<0>(users[i]);
+      std::string user_shares = std::get<1>(users[i]);
+      hierarchy.append(indent + " " + bank_name + "|" + username + "|" + user_shares + "\n");
+    }
+  }
+  // otherwise, this bank has sub banks, so call this helper
+  // function again with the first sub bank it found
+  else {
+    while (exit == SQLITE_ROW) {
+      // execute SQL statement
+      std::string bank = reinterpret_cast<char const*> (sqlite3_column_text(b_select_sub_banks_stmt, 0));
+      banks.push_back(bank);
+      exit = sqlite3_step(b_select_sub_banks_stmt);
+    }
+    for (std::string b : banks) {
+      // reset the prepared statements back to their initial state and
+      // clear their bindings
+      sqlite3_clear_bindings(b_select_associations_stmt);
+      sqlite3_reset(b_select_associations_stmt);
+      sqlite3_clear_bindings(b_select_sub_banks_stmt);
+      sqlite3_reset(b_select_sub_banks_stmt);
+      sqlite3_clear_bindings(b_select_shares_stmt);
+      sqlite3_reset(b_select_shares_stmt);
+      get_sub_banks(DB, b, indent + " ");
+    }
+  }
+
+  return 0;
+}
+
+
+/*
+print_hierarchy.cpp takes one argument, the path to a flux-accounting DB file.
+
+It will look for a parent bank in the bank table. It will exit if it does not
+find a root bank, or if it finds more than one root bank. Once it finds a root
+bank, it will call get_sub_banks, which is a recursive function that traverses
+the bank table with a depth-first search. Once it finds a bank with no sub
+banks, it will print any associations and their corresponding shares under that
+bank.
+*/
+int main(int argc, char** argv) {
```

Comment:
Depending on how we plan on using this code, I think it might make sense to have `main` just do argument parsing, and then put the sqlite opening, prepare, finalize, etc code into a separate function that is public-facing and easily called by other parts of the code base.

As a strawman, the function declarations in this file could be something like:
```
int main(int argc, char** argv);
static int get_sub_banks(sqlite3* DB, const std::string& bank_name, const std::string& indent, sqlite3_stmt *b_select_shares_stmt, sqlite3_stmt *b_select_sub_banks_stmt, sqlite3_stmt *b_select_associations_stmt);
int print_banks(char *db_path);
```

Where `print_banks` would be the public-facing API and `get_sub_banks`/`main` are "internal" functions.

That being said, I think our plan is to expose a public API that populates @dongahn's fair tree data structure, and then the printing is just a walker on that populated tree.  So maybe the public API should be something like `weighted_tree_node_t* populate_banks(char *db_path)`.  Something to discuss in more detail in our 3pm call.